cloudflare_worker/worker/src/index.ts

```diff
@@ -242,11 +279,131 @@ async function handleRequest(request: Request) {
   }
 }
 
+const PAYLOAD_SIZE_LIMIT = 8000000;
+
+// https://datatracker.ietf.org/doc/html/rfc7230#section-3.2.6
+const TOKEN = /^[!#$%&'*+.^_`|~0-9a-zA-Z-]+$/;
+
+// Matcher for HTML with either UTF-8 or unspecified character encoding.
+// Capture group 1 indicates that charset=utf-8 was explicitly stated.
+//
+// https://datatracker.ietf.org/doc/html/rfc7231#section-3.1.1.5
+//
+// The list of aliases for UTF-8 is codified in
+// https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/platform/wtf/text/text_codec_utf8.cc;l=52-68;drc=984c3018ecb2ff818e900fdb7c743fc00caf7efe
+// and https://encoding.spec.whatwg.org/#concept-encoding-get.
+// These are not currently supported, but could be if desired.
+const HTML = /^text\/html([ \t]*;[ \t]*charset=(utf-8|"utf-8"))?$/i;
+
+// If any <link rel=preload>s are found, they are promoted to Link headers.
+// Later, generateSxgResponse will further modify the link header to support SXG
+// preloading of eligible subresources.
+//
+// Out of an abundance of caution, this is limited to documents that are
+// explicitly labeled as UTF-8 via Content-Type or <meta>. This could be
+// expanded in the future, as the risk of misinterpreting type or encoding is
+// rare and low-impact: producing `Link: rel=preload` headers for incorrect
+// refs, which would waste bytes.
+async function promoteLinkTagsToHeaders(payload: Response): Promise<Response> {
+  if (!payload.body) {
+    return payload;
+  }
+
+  let known_utf8 = false;
+
+  // Only run HTMLRewriter if the content is HTML.
+  const content_type_match = payload.headers.get('content-type')?.match(HTML);
+  if (!content_type_match) {
+    return payload;
+  }
+  if (content_type_match[1]) {
+    known_utf8 = true;
+  }
+
+  // A temporary response.
+  let toConsume;
+
+  // Check for UTF-16 BOM, which overrides the <meta> tag, per the implementation at
+  // https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/html/parser/text_resource_decoder.cc;l=394;drc=7a0b88f6d5c015fd3c280b58c7a99d8e1dca28ac
+  // and the spec at
+  // https://html.spec.whatwg.org/multipage/parsing.html#encoding-sniffing-algorithm.
+  [payload, toConsume] = teeResponse(payload);
+  const bom = await readArrayPrefix(toConsume.body, 2);
+  toConsume.body?.cancel();
```

Comment:
Ah, good catch. I was confused because [this example](https://developer.mozilla.org/en-US/docs/Web/API/ReadableStreamDefaultReader/cancel#examples) cancels a reader after it is done. However, the definitions of [`read()`](https://streams.spec.whatwg.org/#default-reader-read) and [`cancel()`](https://streams.spec.whatwg.org/#readable-stream-cancel) imply this is unnecessary. This is also true of the `consumeBytes` call.